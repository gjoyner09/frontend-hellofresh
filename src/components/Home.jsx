export const Home = () => {
    return (
        <>
            <h1>Home</h1>
            <p>Welcome to my Hello Fresh App! Click the links above to see the weekly menus and recipes. Bon Apetit!</p>
        </>
    )
}