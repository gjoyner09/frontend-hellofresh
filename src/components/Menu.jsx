import { useEffect, useState } from "react"
import { api } from "../data"

export const Menu = (props) => {
    const id = props.match.params.id
    const [recipeNumbers, setRecipeNumbers] = useState([])
    const [recipes, setRecipes] = useState([])
    const [allRecipes, setAllRecipes] = useState([])
    const [newRecipe, setNewRecipe] = useState({})
    const [menu, setMenu] = useState({})
    const [errorMessage, setErrorMessage] = useState("")
    let weekRecipes
    
    useEffect(() => {
        api.get("/recipe_week_join_tables")
        .then(({ data }) => {
            weekRecipes = data.filter(({ menu_id }) => menu_id == id)
            console.log("recipe numbers")
            console.log(weekRecipes)
            setRecipeNumbers(weekRecipes)
        })
        .catch(({ message }) => setErrorMessage(`oops, something went wrong: ${message}`))
        
        api.get("/recipes")
        .then(({ data }) => {
            setAllRecipes(data)
            console.log("all recipes")
            console.log(data)
        })
        
        api.get(`/menus/${id}`)
        .then(({ data }) => {
            setMenu(data)
        })
    }, [])
    
    useEffect(() => {
        if (recipeNumbers.length > 0 && allRecipes.length > 0) {
            const matchedRecipes = []
            allRecipes.forEach(recipe => {
                recipeNumbers.forEach(match => {
                    if (recipe.id === match.recipe_id) {
                        matchedRecipes.push(recipe)
                    }
                })
            })
            console.log("matched recipes")
            console.log(matchedRecipes)
            setRecipes(matchedRecipes)
        }
    }, [recipeNumbers, allRecipes])
    
    const addRecipe = (event) => {
        event.preventDefault()
        api.post('/recipe_week_join_tables', {
            menu_id: Number(id),
            recipe_id: Number(newRecipe)
        })
        let recipesCopy = [...recipes]
        const recipeToAdd = allRecipes.filter(recipe => recipe.id === Number(newRecipe) )
        console.log("recipe to add")
        console.log(recipeToAdd)
        recipesCopy.push(recipeToAdd[0])
        console.log(recipesCopy)
        setRecipes(recipesCopy)
        setNewRecipe({})
    }
    
    const updateNewRecipe = (event) => {
        setNewRecipe(event.target.value)
    }
    
    return (
        <>
            <h1>Week of {menu.week}</h1>
            {recipes.map(({title, ingredients, instructions, nutritional_info, classification, id}) => {
                return (
                    <div key={id}>
                        <h2 id={id}>{title}</h2>
                        <ul>
                            <li>Ingredients: {ingredients}</li>
                            <li>Instructions: {instructions}</li>
                            <li>Nutritional Info: {nutritional_info}</li>
                            <li>Classification: {classification}</li>
                        </ul>
                    </div>
                )
            })}
            <form onSubmit={addRecipe}>
                <select name="recipe" id="recipe" value={newRecipe} onChange={updateNewRecipe}>
                    <option></option>
                    {allRecipes.map((recipe) => {
                        return (
                            <option key={recipe.id} value={recipe.id}>{recipe.title}</option>
                        )
                    })}
                </select>
                <button type="submit">Add Recipe</button>
            </form>
        </>
    )
}