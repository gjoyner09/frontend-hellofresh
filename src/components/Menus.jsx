import { useEffect, useState } from "react"
import { api } from "../data"
import { Link } from "react-router-dom"

export const Menus = () => {
    const [menus, setMenus] = useState([])
    const [errorMessage, setErrorMessage] = useState("")
    const [week, setWeek] = useState("")
    
    const addMenu = async (event) => {
        event.preventDefault()
        
        try {
            const { data } = await api.post("/menus", { week })
            setWeek("")
            const menusClone = [...menus]
            menusClone.push({ week: data.week })
            setMenus(menusClone)
        } catch ({ message }) {
            setErrorMessage(`oops, something went wrong: ${message}`)
        }
    }
    
    useEffect(() => {
        api.get("/menus")
        .then(({ data }) => {
            setMenus(data)
        })
        .catch(({ message }) => setErrorMessage(`oops, something went wrong: ${message}`))
    }, [])
    
    return (
        <>
            <h1>Menus</h1>
            {menus.map(({week, id}) => {
                return (
                    <>
                        <Link key={id} to={`/menu/${id}`}>Week of {week}</Link>
                        <br></br>
                    </>
                )
            })}
            <br></br>
            <form onSubmit={addMenu}>
                <input type="text" onChange={(e) => setWeek(e.target.value)} value={week} id="week" label="Week" />
                <label htmlFor="week">Week (YYYY-MM-DD format)</label>
                <br></br>
                <button type="submit">Add menu</button>
            </form>
        </>
    )
}