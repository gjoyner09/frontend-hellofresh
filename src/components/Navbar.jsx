import {
    Link,
    useLocation
  } from "react-router-dom"
  
  
  const navLinks = [
    {
      label: "Home",
      value: "/"
    },
    {
      label: "Menus",
      value: "/menus"
    },
    {
      label: "Recipes",
      value: "/recipes"
    },
    {
      label: "Sign in",
      value: "/signin"
    },
    {
      label: "Sign up",
      value: "/signup"
    }
  ]
  
  export const Navbar = () => {
  
    return (
      <nav position="sticky">
          <h1>MealHub</h1>
            {navLinks.map(({ label, value }) => (
                <div key={value}>
                    <Link label={label} value={value} to={value}>{label}</Link>
                    <br></br>
                </div>
            ))}
        </nav>
    )
  }