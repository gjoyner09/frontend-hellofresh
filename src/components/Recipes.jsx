import { useEffect, useState } from "react"
import { api } from "../data"

class AllRecipes {
    static all() {
      return api.get('/recipes').then(resp => resp.data);
    }
  }
  
export default AllRecipes;

export const Recipes = () => {
    const [recipes, setRecipes] = useState([])
    const [errorMessage, setErrorMessage] = useState("")
    const [title, setTitle] = useState("")
    const [ingredients, setIngredients] = useState("")
    const [instructions, setInstructions] = useState("")
    const [nutritional_info, setNutritionalInfo] = useState("")
    const [classification, setClassification] = useState("")
    
    useEffect(() => {
        api.get("/recipes")
            .then(({ data }) => {
                setRecipes(data)
            })
            .catch(({ message }) => setErrorMessage(`oops, something went wrong: ${message}`))
    }, [])
    
    const addRecipe = async (event) => {
        event.preventDefault()
        try {
          // send a post request to add the new card to the backend
          const { data } = await api.post("/recipes", {
            title,
            ingredients,
            instructions,
            nutritional_info,
            classification
          })
    
          setTitle("")
          setIngredients("")
          setInstructions("")
          setNutritionalInfo("")
          setClassification("")
    
          // update the component state with the new card
          // make a clone of cards from state
          const recipesClone = [...recipes]
          // add the new card
          recipesClone.push({
            title: data.title,
            ingredients: data.ingredients,
            instructions: data.instructions,
            nutritional_info: data.nutritional_info,
            classification: data.classification,
            id: data.id
          })
          // set it as the new state
          setRecipes(recipesClone)
        } catch ({ message }) {
          setErrorMessage(`oops, something went wrong: ${message}`)
        }
      }
    
    return (
        <>
            <h1>Recipes</h1>
            {recipes.map(({id, title, ingredients, instructions, nutritional_info, classification}) => {
                return (
                    <div id={id}>
                        <h1>{title}</h1>
                        <ul>
                            <li>Ingredients: {ingredients}</li>
                            <li>Instructions: {instructions}</li>
                            <li>Nutritional Info: {nutritional_info}</li>
                            <li>Classification: {classification}</li>
                        </ul>
                    </div>
                )
            })}
            <form onSubmit={addRecipe}>
                <input type="text" onChange={(e) => setTitle(e.target.value)} value={title} id="title" label="Title" />
                <label htmlFor="title">Title</label>
                <br></br>
                <input type="text" onChange={(e) => setIngredients(e.target.value)} value={ingredients} id="ingredients" label="Ingedients" />
                <label htmlFor="ingredients">Ingredients</label>
                <br></br>
                <input type="text" onChange={(e) => setInstructions(e.target.value)} value={instructions} id="instructions" label="Instructions" />
                <label htmlFor="instructions">Instructions</label>
                <br></br>
                <input type="text" onChange={(e) => setNutritionalInfo(e.target.value)} value={nutritional_info} id="nutritional_info" label="Nutritional Info" />
                <label htmlFor="nutritional_info">Nutritional Info</label>
                <br></br>
                <input type="text" onChange={(e) => setClassification(e.target.value)} value={classification} id="classification" label="Classification" />
                <label htmlFor="classification">Classification</label>
                <br></br>
                <button type="submit">Add recipe</button>
            </form>
        </>
    )
}