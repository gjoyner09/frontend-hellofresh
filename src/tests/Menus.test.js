import { api } from '../data/index';
import renderer from 'react-test-renderer'
import AllMenus from '../components/Menus';
import { Menus } from '../components/Menus'

jest.mock('axios', () => {
  return {
    create: jest.fn(() => ({
      get: jest.fn(),
      interceptors: {
        request: { use: jest.fn(), eject: jest.fn() },
        response: { use: jest.fn(), eject: jest.fn() }
      }
    }))
  }
})


test('should fetch all menus', () => {
  const menus = [{
      week: "2021-07-05"
    }];
  const resp = {data: menus};
  api.get.mockResolvedValue(resp);

  // or you could use the following depending on your use case:
  // axios.get.mockImplementation(() => Promise.resolve(resp))

  return AllMenus.all().then(data => expect(data).toEqual(menus));
})

test('Menus page renders correctly', () => {
    const component = renderer.create(
        <Menus />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
})