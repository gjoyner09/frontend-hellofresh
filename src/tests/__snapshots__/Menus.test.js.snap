// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menus page renders correctly 1`] = `
Array [
  <h1>
    Menus
  </h1>,
  <br />,
  <form
    onSubmit={[Function]}
  >
    <input
      id="week"
      label="Week"
      onChange={[Function]}
      type="text"
      value=""
    />
    <label
      htmlFor="week"
    >
      Week (YYYY-MM-DD format)
    </label>
    <br />
    <button
      type="submit"
    >
      Add menu
    </button>
  </form>,
]
`;
